### PR TITLE
make root a functional component

### PR DESCRIPTION
**Pull request by rianfowler**
#### Which issue(s) this PR fixes:
https://app.shortcut.com/replicated/story/53446/in-helm-managed-mode-user-is-prompted-to-upload-license

#### Special notes for your reviewer:
- this refactors the Root component into a functional component with type definitions for its props and state. 
- this is necessary refactor for introducing hooks and extracting utility functions from the root component 

## Steps to reproduce
n/a

#### Does this PR introduce a user-facing change?
none- refactor 

#### Does this PR require documentation?
none

### CODE REVIEW

File: web/src/Root.tsx [hunk 1]
```diff
@@ -0,0 +1,672 @@
...
+  };
+
+  // TODO: delete if not used
+  // const handleActiveInitSessionCompleted = () => {
+  //   if (Utilities.localStorageEnabled()) {
+  //     localStorage.removeItem(INIT_SESSION_ID_STORAGE_KEY);
+  //   }
+  //   setState({ initSessionId: "" });
+  // };
+
```
Comment (rianfowler) [inline web/src/Root.tsx:192]: will delete later. I'm not sure why this is here but nothing uses it.

File: web/src/Root.tsx [hunk 2]
```diff
@@ -0,0 +1,672 @@
...
+          "failed to get pending apps, unexpected status code",
+          res.status
+        );
+        return;
+      }
+      const response = await res.json();
+      const app = response.app;
+      // TODO: delete if not used
+      // setState({
+      //   pendingApp: app,
```
Comment (rianfowler) [inline web/src/Root.tsx:242]: this `pendingApp` thing doesn't seem to be used anywhere.

File: web/src/Root.tsx [hunk 3]
```diff
@@ -0,0 +1,672 @@
...
+        />
+        <meta httpEquiv="Pragma" content="no-cache" />
+        <meta httpEquiv="Expires" content="0" />
+        {state.appLogo && (
+          <link rel="icon" type="image/png" href={state.appLogo} />
+        )}
+        {state.appBranding?.fontFaces &&
+          state.appBranding.fontFaces.map((fontFace, index) => (
+            <style
+              type="text/css"
```
Comment (rianfowler) [inline web/src/Root.tsx:425]: I removed the `rel` attribute. I was getting a type error and as far as i can tell it's not something that's available on the `<style>` tag. please double check that.

File: web/src/Root.tsx [hunk 4]
```diff
@@ -0,0 +1,672 @@
...
+
+type App = {
+  slug: string;
+  isGitOpsSupported: boolean;
+  isIdentityServiceSupported: boolean;
+  isGeoaxisSupported: boolean;
+  allowSnapshots: boolean;
+};
+type State = {
+  appBranding: AppBranding | null;
```
Comment (rianfowler) [inline web/src/Root.tsx:89]: there was something called `appBrandingCss` that was initialized but never used. i assume this is the intended name

File: web/src/Root.tsx [hunk 5]
```diff
@@ -0,0 +1,672 @@
...
+  rootDidInitialWatchFetch: boolean;
+  selectedAppName: string | null;
+  snapshotInProgressApps: string[];
+  themeState: ThemeState;
+};
+
+let interval: ReturnType<typeof setInterval> | undefined;
+
+const Root = () => {
+  const [state, setState] = useReducer(
```
Comment (rianfowler) [inline web/src/Root.tsx:112]: following @alicenstar 's suggestion here. this is a lot cleaner to use for converting to stateless components

File: web/src/Root.tsx [hunk 6]
```diff
@@ -0,0 +1,672 @@
...
+    (currentState: State, newState: Partial<State>) => ({
```
Comment (rianfowler) [inline web/src/Root.tsx:113]: learned about `Partial` today

File: web/src/Root.tsx [hunk 7]
```diff
@@ -0,0 +1,672 @@
...
+     * Reference object to a blank theme state
+     */
+    const EMPTY_THEME_STATE = {
+      navbarLogo: null,
+    };
+
+    setState({
+      themeState: { ...EMPTY_THEME_STATE },
+    });
+  };
```
[the remark this answers is not in the document]
Comment (miaawong) [reply web/src/Root.tsx:176]: great idea!

File: web/src/Root.tsx (hunk 2 above)
Comment (miaawong) [reply web/src/Root.tsx:242]: for future: can checkIsHelmManage be a custom hook? I've seen it used in a few places

File: web/src/Root.tsx (hunk 3 above)
Comment (miaawong) [reply web/src/Root.tsx:425]: I think rel is not necessary on the style tag

**Review by miaawong — APPROVED** [review]
looks great!

File: web/src/Root.tsx (hunk 2 above)
Comment (rianfowler) [reply web/src/Root.tsx:242]: yeah there's a hook for it now. going to move it out of the root component and put it into a router and have it check the same way auth will work.

**Pull request merged**